### PR TITLE
State __sizeof__ impl

### PR DESCRIPTION
**Pull request by cjmartian**
Signed-off-by: Connor Martin <connormartin7@gmail.com>

## Summary

Find size of `State` object and it's children as requested in https://github.com/PrefectHQ/prefect/issues/5230

## Changes

Adds a `__sizeof__` implementation to `State`

## Importance

Was requested in https://github.com/PrefectHQ/prefect/issues/5230 as to correctly report the size of state.

## Checklist

This PR:

- [ ] adds new tests (if appropriate)
- [x] adds a change file in the `changes/` directory (if appropriate)
- [X] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

### CODE REVIEW

**Comment by madkinsz** [comment]
Hey @cjmartian — sorry for the delayed review here!

This is a sweet implementation :) but I think collecting the deep size is probably overkill. We should able to just include the size of `State.result` without recursively descending. Is there a reason you felt a recursive implementation was necessary?

**Comment by cjmartian** [comment]
Hey @madkinsz,

Whoops, I just figured the `State` class should get the complete size of the `State` object, instead of just result. Maybe another method that just returns size of the result?

```
def get_result_size(self):
  return sys.getsizeof(self.result)
```

**Comment by madkinsz** [comment]
Generally `__sizeof__` implementations in Python do not retrieve deep sizes. Here, we want to include the size of the `result` so that Dask can optimizing passing states over the network so it'd have to be implemented in the dunder. I'm actually not sure if the `result` size will be sufficient for Dask but it seems like a good first step. Note that we want to include the overhead of the `State` object still so we'll need to use something like `super().__sizeof__() + getsizeof(self.result)`

File: src/prefect/engine/state.py [hunk 1]
```diff
@@ -110,6 +111,9 @@ def __getstate__(self) -> dict:
 
         return data
 
+    def __sizeof__(self) -> int:
+        return super().__sizeof__() + sys.getsizeof(self.restult)
```
[the remark this answers is not in the document]
Comment (cjmartian) [reply src/prefect/engine/state.py]: Oops, just saw this. Thanks for catching it too!

**Comment by madkinsz** [comment]
@cjmartian thanks! Can you add a test case covering this and update the changes entry?

**Comment by cjmartian** [comment]
@madkinsz Let me know about that test, want to make sure it's actually useful.

File: tests/engine/test_state.py [hunk 2]
```diff
@@ -749,3 +752,14 @@ def __init__(self, *args) -> None:
...
+
+@patch("prefect.engine.state.super")
+def test_state_sizeof(mock_super):
+    state = State(result=10)
+    mock_super().__sizeof__.return_value = 55
+    # Don't use sys.getsizeof(state) here because
+    # getsizeof() includes overhead that is hard
+    # to account for.
+    # Instead call __sizeof__ directly
+    assert state.__sizeof__() == 55 + sys.getsizeof(state.result)
```
Comment (madkinsz) [inline tests/engine/test_state.py:765]: This LGTM but I'll have @abrookins take a look too.

Perhaps one additional assertion that `sys.getsizeof(state) >= state.__sizeof__()` so we can be sure it's being applied in practice?

Comment (cjmartian) [reply tests/engine/test_state.py:765]: Sure thing, I can add that to the test

**Review by madkinsz — APPROVED** [review]

**Comment by madkinsz** [comment]
Thanks!

File: tests/engine/test_state.py (hunk 2 above)
Comment (abrookins) [reply tests/engine/test_state.py:765]: I declare retroactively that this looks good to me. 😄

**Pull request merged**